src/crlf.c

```diff
@@ -146,16 +149,26 @@ static int crlf_apply_to_odb(
 		if (stats.cr != stats.crlf)
 			return GIT_PASSTHROUGH;
 
-		/* If there are no CR characters to filter out and CrLf is not set to "true", then just pass */
-		if (!stats.cr && ca->auto_crlf != GIT_AUTO_CRLF_TRUE)
-			return GIT_PASSTHROUGH;
+		if (ca->crlf_action == GIT_CRLF_GUESS) {
+			/*
+			 * If the file in the index has any CR in it, do not convert.
+			 * This is the new safer autocrlf handling.
+			 */
+			if (has_cr_in_index(src))
+				return GIT_PASSTHROUGH;
+		}
+	}
 
-		/* If safecrlf is enabled, sanity-check the result. */
-		if (stats.lf != stats.crlf) {
+	/* If safecrlf is enabled, sanity-check the result. */
+	if (ca->crlf_action == GIT_CRLF_INPUT ||
+		(ca->auto_crlf == GIT_AUTO_CRLF_INPUT &&
+		(ca->crlf_action == GIT_CRLF_GUESS || ca->crlf_action == GIT_CRLF_AUTO ||
+		(ca->crlf_action == GIT_CRLF_TEXT && ca->eol==GIT_EOL_UNSET)))) {
```

Comment:
nit: missing spaces around the last ==.

But more importantly, this is another case of "sanity-checking" that doesn't make clear what exactly is being checked and which is an if statement with a thousand clauses. This one is somewhat less heinous than the other one, but it does seem to cover a lot of the same ground, so I wonder if you missed some cases here, or if we could extract the common logic between the two, or something.